hbase-server/src/main/java/org/apache/hadoop/hbase/ipc/AbstractBatchDecoder.java

```diff
@@ -0,0 +1,442 @@
+ *               // batch submit to an executor
+ *               for (Object m : (List<?>) msg) {
+ *                   RpcCommandHandler.this.process(ctx, m);
+ *               }
+ *           }
+ *       });
+ *   } else {
+ *       process(ctx, msg);
+ *   }
+ * </pre>
+ * You can check the method {@link AbstractBatchDecoder#channelRead(ChannelHandlerContext, Object)} ()}
+ *   to know the detail modification.
+ */
+public abstract class AbstractBatchDecoder extends ChannelInboundHandlerAdapter {
+    /**
+     * Cumulate {@link ByteBuf}s by merge them into one {@link ByteBuf}'s, using memory copies.
+     */
+    public static final Cumulator MERGE_CUMULATOR     = new Cumulator() {
+                                                          @Override
+                                                          public ByteBuf cumulate(ByteBufAllocator alloc,
+                                                                                  ByteBuf cumulation,
+                                                                                  ByteBuf in) {
+                                                              ByteBuf buffer;
+                                                              if (cumulation.writerIndex() > cumulation
+                                                                  .maxCapacity()
+                                                                                             - in.readableBytes()
+                                                                  || cumulation.refCnt() > 1) {
+                                                                  // Expand cumulation (by replace it) when either there is not more room in the buffer
+                                                                  // or if the refCnt is greater then 1 which may happen when the user use slice().retain() or
+                                                                  // duplicate().retain().
+                                                                  //
+                                                                  // See:
+                                                                  // - https://github.com/netty/netty/issues/2327
+                                                                  // - https://github.com/netty/netty/issues/1764
+                                                                  buffer = expandCumulation(alloc,
+                                                                      cumulation,
+                                                                      in.readableBytes());
+                                                              } else {
+                                                                  buffer = cumulation;
+                                                              }
+                                                              buffer.writeBytes(in);
+                                                              in.release();
```

Comment:
This is all shifted to the right.

Reply:
@saintstack I have already reformat the whole code of this class.